### PR TITLE
PMM-6389 Add VictoriaMetrics

### PR DESCRIPTION
**Pull request by ctrlok**
- [ ] https://github.com/Percona-Lab/percona-images/pull/86
- [ ] https://github.com/percona/pmm-update/pull/96
- [ ] https://github.com/percona/pmm-managed/pull/447

https://github.com/Percona-Lab/pmm-submodules/pull/1020

### CODE REVIEW

File: services/agents/registry.go [hunk 1]
```diff
@@ -62,9 +62,9 @@ type pmmAgentInfo struct {
 //
 // TODO Split into several types https://jira.percona.com/browse/PMM-4932
 type Registry struct {
-	db         *reform.DB
-	prometheus prometheusService
-	qanClient  qanClient
+	db             *reform.DB
+	scrapeServices []prometheusService
```
Comment (BupycHuk) [inline services/agents/registry.go:66]: isn't it better to create some proxy service which will pass request to prometheus service or victoria metrics service or both based on environment variables?

File: services/agents/registry.go [hunk 2]
```diff
@@ -392,8 +392,9 @@ func (r *Registry) stateChanged(ctx context.Context, req *agentpb.StateChangedRe
 	if e != nil {
 		return e
 	}
-
-	r.prometheus.RequestConfigurationUpdate()
+	for _, svc := range r.scrapeServices {
```
Comment (BupycHuk) [inline services/agents/registry.go:395]: So we will not have to write these loops everywhere.

File: services/agents/registry.go (hunk 1 above)
Comment (BupycHuk) [reply services/agents/registry.go:66]: And then remove prometheus will be easier.

File: services/prometheus/scrape_configs.go [hunk 3]
```diff
@@ -474,3 +477,168 @@ func scrapeConfigsForExternalExporter(s *models.MetricsResolutions, params *scra
 
 	return []*config.ScrapeConfig{cfg}, nil
 }
+
+// PopulateScrapeConfigs populates scrape configs for basic running service and agents.
+func PopulateScrapeConfigs(cfg *config.Config, l *logrus.Entry, q *reform.Querier, s *models.MetricsResolutions) error {
```
Comment (BupycHuk) [inline services/prometheus/scrape_configs.go:482]: Is it a good idea to make it public? Maybe it's better to create separate structure which will handle all scraper_configs and pass it, so victoriaMetrics packaged will not use prometheus package. or move this file somewhere else.
It should help us with removing prometheus once victoria will be ready.

**Review by BupycHuk — CHANGES_REQUESTED** [review]
CI Failed

File: services/supervisord/supervisord.go [hunk 4]
```diff
@@ -558,3 +570,28 @@ stdout_logfile_backups = 3
...
+priority = 7
+command =
+	/usr/sbin/victoria-metrics
+		--promscrape.config=/etc/victoriametrics-promscrape.yml
+		--retentionPeriod={{ .DataRetentionMonth }}
+		--storageDataPath=/srv/victoriametrics/data
+		--httpListenAddr=127.0.0.1:8428
+user = pmm
+autorestart = true
+autostart = true
```
Comment (AlekSi) [inline services/supervisord/supervisord.go:587]: We should render that file in all cases, but that value, and one above, should be `false` if `isVMEnabled` is false

**Review by AlekSi — CHANGES_REQUESTED** [review]

File: services/victoriametrics/victoriametrics.go [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+	"gopkg.in/yaml.v2"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/services/prometheus"
+)
+
+const updateBatchDelay = 3 * time.Second
+
+var (
+	enabled       = os.Getenv("PERCONA_TEST_VM") == "1"
```
Comment (AlekSi) [inline services/victoriametrics/victoriametrics.go:48]: And what if it is `"true"`?

File: services/agents/registry.go (hunk 1 above)
Comment (tenmozes) [reply services/agents/registry.go:66]: we don't need to do orchestration for removing of prometheus, i think this is not a hard work. 
Please, also take into account that `vmalert` will be added and it has similar interface and should be reload by when config changes, as right now, this is one call (change settings) we need to ask every service about config verification/updating

File: services/prometheus/scrape_configs.go (hunk 3 above)
Comment (f41gh7) [reply services/prometheus/scrape_configs.go:482]: I think, that removing prometheus will require some work anyway and copy paste code isn't good. For now we can reuse it.

File: services/victoriametrics/victoriametrics.go (hunk 5 above)
Comment (f41gh7) [reply services/victoriametrics/victoriametrics.go]: Added true case.

File: services/supervisord/supervisord.go (hunk 4 above)
Comment (f41gh7) [reply services/supervisord/supervisord.go]: Done, config rendered in all cases.

File: services/prometheus/scrape_configs.go [hunk 6]
```diff
@@ -474,3 +477,168 @@ func scrapeConfigsForExternalExporter(s *models.MetricsResolutions, params *scra
...
+			}
+			paramsHost = externalExporterNode.Address
+		default:
+			l.Warnf("It's not possible to get host, skipping scrape config for %s.", agent)
+
+			continue
+		}
+
+		var scfgs []*config.ScrapeConfig
+		switch agent.AgentType {
```
[the remark this answers is not in the document]
Comment (f41gh7) [reply services/prometheus/scrape_configs.go:551]: checked above if agent.AgentType == models.PMMAgentType {

File: services/victoriametrics/victoriametrics.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+	"gopkg.in/yaml.v2"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/services/prometheus"
+)
+
+const updateBatchDelay = 3 * time.Second
+
+var (
+	enabled       = os.Getenv("PERCONA_TEST_VM") == "1" || os.Getenv("PERCONA_TEST_VM") == "true"
```
Comment (AlekSi) [inline services/victoriametrics/victoriametrics.go:48]: And what if it is "TRUE"? Please use strconv.ParseBool as in other pmm-managed code

Comment (f41gh7) [reply services/victoriametrics/victoriametrics.go]: change dto strconv.Bool

**Pull request closed without merging**